clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -200,6 +200,41 @@ bool Sema::isKnownGoodSYCLDecl(const Decl *D) {
   return false;
 }
 
+bool isZeroSizedArray(QualType Ty) {
+  if (const auto *CATy = dyn_cast<ConstantArrayType>(Ty)) {
+    return (CATy->getSize() == 0);
+  }
+  return false;
+}
+
+void Sema::checkSYCLVarDeclIfInKernel(VarDecl *Var) {
+  // not all variable types supported in kernel contexts
+  // if not we record a deferred diagnostic.
+  assert(getLangOpts().SYCLIsDevice &&
+         "Should only be called during SYCL compilation");
+  QualType Ty = Var->getType();
+  SourceRange Loc = Var->getLocation();
+
+  // __int128, __int128_t, __uint128_t
+  if (Ty->isSpecificBuiltinType(BuiltinType::Int128) ||
+      Ty->isSpecificBuiltinType(BuiltinType::UInt128))
+    SYCLDiagIfDeviceCode(Loc.getBegin(), diag::err_type_unsupported)
+        << Ty.getUnqualifiedType().getCanonicalType().getAsString();
+
+  // QuadType __float128
+  if (Ty->isSpecificBuiltinType(BuiltinType::Float128) &&
+      !Context.getTargetInfo().hasFloat128Type())
+    SYCLDiagIfDeviceCode(Loc.getBegin(), diag::err_type_unsupported)
+        << "__float128";
+
+  // zero length arrays
+  if (isZeroSizedArray(Ty))
+    SYCLDiagIfDeviceCode(Loc.getBegin(), diag::err_typecheck_zero_array_size);
+
+  // TODO: check type of accessor
```

Comment:
??

Reply:
BTW I have a feeling that it is not a proper place for accessor type checking. Just because accessors always declared in host code and captured to the device code. We can instead:

- Check type of buffer/accessor data using static assertions in SYCL headers.
- Check type of accessors while we iterate over kernel object captures in SemaSYCL (where check for trivially destructible/constructible kernel arguments here).

I personally prefer the way with SYCL headers, since SemaSYCL extremely big, complicated and ugly.

Reply:
Ping?

Reply:
I'll remove the comment. This is the next thing I am planning to undertake, but not as part of this PR.